From 66fb9f13289c0c1c29fb43942ce3ba69dbce5916 Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@gmail.com>
Date: Sun, 13 Nov 2022 11:20:08 -0800
Subject: [PATCH 183/196] Update minerfund addresses for numbers upgrade (#285)

Use new minerfund addresses for numbers upgrade and change the way
they're handled so that they select only one output, iteratively, per
block. This avoids having tons of tiny tiny outputs.
---
 src/consensus/addresses.h                     | 16 +++++
 src/consensus/params.h                        |  1 +
 src/minerfund.cpp                             | 55 ++++++++++-----
 .../logos_feature_minerfund_activation.py     | 67 ++++++++++++++++++-
 4 files changed, 119 insertions(+), 20 deletions(-)

diff --git a/src/consensus/addresses.h b/src/consensus/addresses.h
index 8ee04bb2d..235eadafb 100644
--- a/src/consensus/addresses.h
+++ b/src/consensus/addresses.h
@@ -92,5 +92,21 @@ Consensus::CoinbaseAddresses AddressSets = {
             // Lotus Citizen 4
             "lotus_16PSJNEqzEoLhBigDN8Hf93FzERrPQKJXaYdzRCNT",
         },
+    .numbers =
+        {
+            "lotus_16PSJKzohkeX3G5BKpbeESKZZCHWrQymNRwjsu42Y",
+            "lotus_16PSJHtDqN7ibmc4NSziqbrmAXoSLE46F9LceMUDx",
+            "lotus_16PSJHhULzYgupzGnvDysUFLxYxBVwx92JjXSyjNL",
+            "lotus_16PSJLTMsACkicNJVmAPBQgmEeagvkXKuF8gR2qE8",
+            "lotus_16PSJPfz8J1KkScQUMZK8GabJgTGxr8c5WWidEQYS",
+            "lotus_16PSJN2GzCj6PVwg4oTbiXftP7Jf3Ux2zB3X4A6Dw",
+            "lotus_16PSJMtQgty32UmgzkomZYPavZaJb8kn3HUeD4jno",
+            "lotus_16PSJLMtCuFnyMb3rz4zbbhAtJgppp8GiY5NpGGev",
+            "lotus_16PSJP1z4qX4XeJprHPgw2GbY86GxMwT2DKmVmHqS",
+            "lotus_16PSJNob8nN2U5e1WfpvpHbn28iJ8mr5ZbjsTNmzk",
+            "lotus_16PSJM1j9nPx6xLbyJ4UMsT1XNMdZAsT6vayCiEhh",
+            "lotus_16PSJL7B1pQm4pT3M4k6BeXNR8PXkvuv8BeosgUdA",
+            "lotus_16PSJNfvYdAquq6X28omdnpMyixZX9JfF9MgmDX3H",
+        },
 };
 } // namespace RewardAddresses
diff --git a/src/consensus/params.h b/src/consensus/params.h
index ba6b4b772..f8540d7e7 100644
--- a/src/consensus/params.h
+++ b/src/consensus/params.h
@@ -18,6 +18,7 @@ struct CoinbaseAddresses {
     std::vector<std::string> genesis;
     std::vector<std::string> exodus;
     std::vector<std::string> leviticus;
+    std::vector<std::string> numbers;
 };
 
 /**
diff --git a/src/minerfund.cpp b/src/minerfund.cpp
index b9dc7731a..5bd614625 100644
--- a/src/minerfund.cpp
+++ b/src/minerfund.cpp
@@ -5,6 +5,7 @@
 #include <minerfund.h>
 
 #include <cashaddrenc.h> // For DecodeCashAddrContent
+#include <chain.h>       // For CBlockIndex class
 #include <chainparams.h>
 #include <consensus/activation.h>
 #include <key_io.h> // For DecodeDestination
@@ -23,16 +24,28 @@ static const CTxOut BuildOutput(const std::string &address,
     return {amount, script};
 }
 
-static const std::vector<std::string> &
-GetPayoutAddresses(const Consensus::Params &params,
-                   const CBlockIndex *pindexPrev) {
-    if (IsLeviticusEnabled(params, pindexPrev)) {
-        return params.coinbasePayoutAddresses.leviticus;
-    }
-    if (IsExodusEnabled(params, pindexPrev)) {
-        return params.coinbasePayoutAddresses.exodus;
+static const std::vector<CTxOut>
+BuildOutputsCycling(const std::vector<std::string> &addresses,
+                    const CBlockIndex *pindexPrev, const Amount blockReward) {
+    const size_t numAddresses = addresses.size();
+    const Amount shareAmount = blockReward / 2;
+    const auto blockHeight = pindexPrev->nHeight + 1;
+    const auto addressIndx = blockHeight % numAddresses;
+    const auto address = addresses[addressIndx];
+    return std::vector<CTxOut>({BuildOutput(address, shareAmount)});
+}
+
+static const std::vector<CTxOut>
+BuildOutputsFanOut(const std::vector<std::string> &addresses,
+                   const CBlockIndex *pindexPrev, const Amount blockReward) {
+    const size_t numAddresses = addresses.size();
+    std::vector<CTxOut> minerFundOutputs;
+    const Amount shareAmount = blockReward / (int64_t(2 * numAddresses));
+    minerFundOutputs.reserve(numAddresses);
+    for (const std::string &address : addresses) {
+        minerFundOutputs.emplace_back(BuildOutput(address, shareAmount));
     }
-    return params.coinbasePayoutAddresses.genesis;
+    return minerFundOutputs;
 }
 
 std::vector<CTxOut> GetMinerFundRequiredOutputs(const Consensus::Params &params,
@@ -43,15 +56,21 @@ std::vector<CTxOut> GetMinerFundRequiredOutputs(const Consensus::Params &params,
         return {};
     }
 
-    const std::vector<std::string> &addresses =
-        GetPayoutAddresses(params, pindexPrev);
-    const size_t numAddresses = addresses.size();
-    std::vector<CTxOut> minerFundOutputs;
-    const Amount shareAmount = blockReward / (int64_t(2 * numAddresses));
-    minerFundOutputs.reserve(numAddresses);
-    for (const std::string &address : addresses) {
-        minerFundOutputs.emplace_back(BuildOutput(address, shareAmount));
+    if (IsNumbersEnabled(params, pindexPrev)) {
+        return BuildOutputsCycling(params.coinbasePayoutAddresses.numbers,
+                                   pindexPrev, blockReward);
     }
 
-    return minerFundOutputs;
+    if (IsLeviticusEnabled(params, pindexPrev)) {
+        return BuildOutputsFanOut(params.coinbasePayoutAddresses.leviticus,
+                                  pindexPrev, blockReward);
+    }
+
+    if (IsExodusEnabled(params, pindexPrev)) {
+        return BuildOutputsFanOut(params.coinbasePayoutAddresses.exodus,
+                                  pindexPrev, blockReward);
+    }
+
+    return BuildOutputsFanOut(params.coinbasePayoutAddresses.genesis,
+                              pindexPrev, blockReward);
 }
diff --git a/test/functional/logos_feature_minerfund_activation.py b/test/functional/logos_feature_minerfund_activation.py
index d9051de32..f5fa36caa 100755
--- a/test/functional/logos_feature_minerfund_activation.py
+++ b/test/functional/logos_feature_minerfund_activation.py
@@ -29,7 +29,7 @@ EXODUS_ACTIVATION_TIME = 2000000000
 LEVITICUS_ACTIVATION_TIME = 2010000000
 NUMBERS_ACTIVATION_TIME = 2020000000
 
-# see consensus/addresses.h
+# see consensus/addresses.h, use getaddressinfo to get the scriptPubKey
 GENESIS_SCRIPTS = [
     "a914b6c79031b71d86ab0d617e1e1e706ec4ee34b07f87",
     "76a914b8ae1c47effb58f72f7bca819fe7fc252f9e852e88ac",
@@ -75,6 +75,22 @@ LEVITICUS_SCRIPTS = [
     "76a91419e8d8d0edab6ec43f04b656bff72af78d63ff6588ac",
     "76a914a77122c592e9d24d51ad1dae37de73d77d7bc7f788ac",
 ]
+NUMBERS_SCRIPTS = [
+    "76a9145e38e8e232aca63350fb59a0d8465189c7d0549188ac",
+    "76a914193354a1a0b2e3097a1d54c2958207fff6ff775188ac",
+    "76a9141325d2d8ba6e8c7d99ff66d21530917cc73429d288ac",
+    "76a9146d2c8cb2d031c09862c5fbde900ae8cc0570d1ce88ac",
+    "76a914d64134ac890bd18831560f0c4ef5cea5c976484888ac",
+    "76a914a05d68eaceaefe5cdc1b0794a762f21cf578c5d688ac",
+    "76a9149bef68961730d891b117424de78f47891b7d8c5f88ac",
+    "76a9146a171891ab9443020bd2755ef79c6e59efc5926588ac",
+    "76a914c0dba2b28f0fb9b49b58135ce37fef8ce978b75888ac",
+    "76a914b9e0eeacd624c507796bce80316eb836c495d2cc88ac",
+    "76a9147f66047ab92de6ad4d62f609fc123c49fb63d4ff88ac",
+    "76a91461cebabeaa6365887bb95d95e62b848fa365506388ac",
+    "76a914b59008144c4f883e3b1fd9e18e479bc2f877fca788ac",
+]
+
 
 class MinerFundActivationTest(BitcoinTestFramework):
     def set_test_params(self):
@@ -106,7 +122,22 @@ class MinerFundActivationTest(BitcoinTestFramework):
                 coinbase.vout.append(CTxOut(int(SUBSIDY * COIN) // 26, CScript(bytes.fromhex(script))))
             coinbase.rehash()
             block = create_block(int(best_block_hash, 16),
-                                coinbase, block_height, block_time + 1)
+                                 coinbase, block_height, block_time + 1)
+            return block
+
+        # Make a block with the appropriate coinbase output post-numbers upgrade
+        def make_block_cb_post_numbers(scripts):
+            best_block_hash = node.getbestblockhash()
+            block_time = node.getblockheader(best_block_hash)['time']
+            block_height = node.getblockcount() + 1
+            coinbase = create_coinbase(block_height)
+            coinbase.vout[1].nValue = int(SUBSIDY * COIN) // 2
+            coinbase.vout[1].scriptPubKey = CScript([OP_HASH160, bytes(20), OP_EQUAL])
+            script = scripts[block_height % len(scripts)]
+            coinbase.vout.append(CTxOut(int(SUBSIDY * COIN) // 2, CScript(bytes.fromhex(script))))
+            coinbase.rehash()
+            block = create_block(int(best_block_hash, 16),
+                                 coinbase, block_height, block_time + 1)
             return block
 
         # Pre-fork minerfund with wrong (i.e. post-fork) scripts
@@ -181,6 +212,38 @@ class MinerFundActivationTest(BitcoinTestFramework):
         prepare_block(block)
         assert_equal(node.submitblock(ToHex(block)), None)
 
+        # Set mocktime for Leviticus activation
+        node.setmocktime(NUMBERS_ACTIVATION_TIME)
+        # Mine 11 blocks with LEVITICUS_ACTIVATION_TIME in the middle
+        # That moves MTP exactly to LEVITICUS_ACTIVATION_TIME
+        for i in range(-6, 6):
+            block = make_block_with_cb_scripts(LEVITICUS_SCRIPTS)
+            block.nTime = NUMBERS_ACTIVATION_TIME + i
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), None)
+        assert_equal(node.getblockchaininfo()['mediantime'], NUMBERS_ACTIVATION_TIME)
+
+        # Now the using the genesis addresses fails
+        block = make_block_with_cb_scripts(GENESIS_SCRIPTS)
+        prepare_block(block)
+        assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+
+        # Now the using the exodus addresses fails
+        block = make_block_with_cb_scripts(EXODUS_SCRIPTS)
+        prepare_block(block)
+        assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+
+        # Using new scripts now works
+        block = make_block_with_cb_scripts(LEVITICUS_SCRIPTS)
+        prepare_block(block)
+        assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+
+        # now mine 26 blocks to ensure all addresses pass
+        for i in range(0, 26):
+            block = make_block_cb_post_numbers(NUMBERS_SCRIPTS)
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), None)
+
 
 if __name__ == '__main__':
     MinerFundActivationTest().main()
-- 
2.39.5 (Apple Git-154)

